src/System.IO.Pipelines.Networking.Tls/ApplicationProtocols.cs

```diff
@@ -0,0 +1,157 @@
+
+        private static readonly byte[] _confidentialWebRTCMediaAndData =
+        {
+            0x63, 0x2d, 0x77, 0x65, 0x62, 0x72, 0x74, 0x63
+        }; //("c-webrtc")
+
+        private static readonly byte[] _Ftp = {0x66, 0x74, 0x70}; //("ftp")
+
+        private static readonly byte[][] _allProtocols =
+        {
+            _http1, _spdy1, _spdy2, _spdy3, _traversalUsingRelaysAroundNAT,
+            _natDiscoveryUsingSessionTraversalUtilitiesforNAT, _http2overTLS, _http2overTCP, _webRTCMediaAndData,
+            _confidentialWebRTCMediaAndData, _Ftp
+        };
+
+        private static readonly Array _numberOfProtocols = Enum.GetValues(typeof(ProtocolIds));
+
+        const int _HeaderLength = 10;
+        const int _ContentOffset = 4;
+
+        [Flags]
+        public enum ProtocolIds
+        {
+            None = 0,
+            Http11 = 1,
+            Spdy1 = 2,
+            Spdy2 = 4,
+            Spdy3 = 8,
+            TraversalUsingRelaysAroundNat = 16,
+            NatDiscoveryUsingSessionTraversalUtilitiesforNat = 32,
+            Http2OverTls = 64,
+            Http2OverTcp = 128,
+            WebRtcMediaAndData = 256,
+            ConfidentialWebRtcMediaAndData = 512,
+            Ftp = 1024
+        }
+
+        internal enum ApplicaitonProtocolNegotiationExtension
+        {
+            None = 0,
+            Npn,
+            Alpn,
+        }
+
+        internal enum ApplicationProtocolNegotiationStatus : uint
+        {
+            None = 0,
+            Success,
+            SelectedClientOnly
+        }
+
+        internal static unsafe ProtocolIds FindNegotiatedProtocol(SSPIHandle context)
+        {
+            ContextApplicationProtocol protoInfo;
+            Interop.QueryContextAttributesW(ref context, ContextAttribute.ApplicationProtocol, out protoInfo);
+
+            if (protoInfo.ProtoNegoStatus != ApplicationProtocolNegotiationStatus.Success)
+            {
+                throw new InvalidOperationException("Could not negotiate a mutal application protocol");
+            }
+            return GetNegotiatedProtocol(protoInfo.ProtocolId, protoInfo.ProtocolIdSize);
+        }
+
+        public static byte[] GetBufferForProtocolId(ProtocolIds supportedProtocols, bool withAlpnHeader)
+        {
+            short listLength = 0;
+            for (int i = 0; i < _numberOfProtocols.Length; i++)
+            {
+                if (((int) _numberOfProtocols.GetValue(i) & (int) supportedProtocols) > 0)
+                {
+                    listLength += (short) (_allProtocols[i].Length + 1);
+                }
+            }
+            byte[] returnValue;
+            if (withAlpnHeader)
+            {
+                //Now we know the total length of the list lets create our array of protocols and header
+                returnValue = new byte[listLength + _HeaderLength];
+                var spa = new Span<byte>(returnValue);
+                int length = _HeaderLength + listLength - _ContentOffset;
```

Comment:
So HeaderLength and ContentOffset are about ALPN?  Perhaps AlpnHeaderLength, AlpnContentOffset would be better names?